Wip/hughsie/synapticsmst ng

hughsie opened the pull request:
Just an idea to reduce complexity. Ideas very welcome.

hughsie commented:
> I'm worried the kernel events won't translate effectively, especially if the topology changed inbetween when fwupd started up and calling the firmware write command.

Would calling `fu_device_rescan()` just before `write_firmware()` fix that issue?

hughsie commented on plugins/synapticsmst/fu-plugin-synapticsmst.c:
```diff
@@ -39,468 +42,109 @@ fu_synapticsmst_check_amdgpu_safe (GError **error)
-		tmp = g_strdup_printf ("test-%s", chip_id_str);
-		fu_plugin_synapticsmst_create_simple_guid (fu_device, device, tmp);
+	/* interesting device? */
+	if (g_strcmp0 (fu_udev_device_get_subsystem (device), "drm_dp_aux_dev") != 0)
```
We use the same function during coldplug at startup, i'm just covering my bases in case that a `drm_dp_aux_dev` class device somehow gets created at runtime.

hughsie commented on plugins/synapticsmst/fu-self-test.c:
```diff
@@ -16,23 +16,47 @@ static void
+		ret = fu_plugin_runner_udev_device_added (plugin, dev, &error);
+		g_assert_no_error (error);
+		g_assert (ret);
+	}
```
Right, it's a lot cleaner.

superm1 commented:
> Would calling fu_device_rescan() just before write_firmware() fix that issue?

Well the problem you run into is that write_firmware () is taking a `FuDevice` as an argument.  if that device is gone, or moved to a different port or cascaded now, is it really still the same `FuDevice`?  And how are you addressing it properly?

It feels to me like you need the lifecycle of the `FuDevice` to be accurate at the time you called the flash command.

hughsie commented:
> is it really still the same FuDevice

Valid point. Perhaps we do a `rescan()` and abort if anything is different compared to before the rescan?

superm1 commented:
> Valid point. Perhaps we do a rescan() and abort if anything is different compared to before the rescan?

I feel like at this point we're bandaging onto the (broken) behavior from the kernel TBH.  Not to say what's there isn't already a bandage of course.  Could you ask the RH GFX guys if they can fix the kernel so we see change events on drm_dp_aux when the topology changes?

hughsie commented:
> I feel like at this point we're bandaging onto the (broken) behavior from the kernel

Yes, I think you're right. As a perhaps half-way house, do you know if we get uevents for the drm class events on unplug/replug? We could proxy those in the plugin until we get a proper kernel event on the actual rm_dp_aux_dev device.

superm1 commented:
I experimented a bit and see some DRM events, but a lot more than "1".  So I think there would need to be some amount of rate limiting or you're going to have 

1) plugging in DP device to type-C port using DP alt mode

```
KERNEL[37.430371] add      /devices/pci0000:00/0000:00:02.0/drm/card0/card0-DP-3 (drm)
KERNEL[37.430553] change   /devices/pci0000:00/0000:00:02.0/drm/card0 (drm)
UDEV  [37.442367] add      /devices/pci0000:00/0000:00:02.0/drm/card0/card0-DP-3 (drm)
UDEV  [37.447759] change   /devices/pci0000:00/0000:00:02.0/drm/card0 (drm)
KERNEL[37.464519] add      /devices/pci0000:00/0000:00:02.0/i2c-10 (i2c)
KERNEL[37.464700] add      /devices/pci0000:00/0000:00:02.0/drm/card0/card0-DP-4 (drm)
KERNEL[37.464773] change   /devices/pci0000:00/0000:00:02.0/drm/card0 (drm)
KERNEL[37.464925] add      /devices/pci0000:00/0000:00:02.0/drm/card0/card0-DP-5 (drm)
KERNEL[37.464983] change   /devices/pci0000:00/0000:00:02.0/drm/card0 (drm)
KERNEL[37.465058] change   /devices/pci0000:00/0000:00:02.0/drm/card0 (drm)
UDEV  [37.468257] add      /devices/pci0000:00/0000:00:02.0/i2c-10 (i2c)
KERNEL[37.475976] change   /devices/pci0000:00/0000:00:02.0/drm/card0 (drm)
KERNEL[37.477399] change   /devices/pci0000:00/0000:00:02.0/drm/card0 (drm)
UDEV  [37.478411] add      /devices/pci0000:00/0000:00:02.0/drm/card0/card0-DP-4 (drm)
UDEV  [37.483032] change   /devices/pci0000:00/0000:00:02.0/drm/card0 (drm)
UDEV  [37.492451] add      /devices/pci0000:00/0000:00:02.0/drm/card0/card0-DP-5 (drm)
UDEV  [37.497025] change   /devices/pci0000:00/0000:00:02.0/drm/card0 (drm)
UDEV  [37.500627] change   /devices/pci0000:00/0000:00:02.0/drm/card0 (drm)
UDEV  [37.503313] change   /devices/pci0000:00/0000:00:02.0/drm/card0 (drm)
UDEV  [37.505537] change   /devices/pci0000:00/0000:00:02.0/drm/card0 (drm)
KERNEL[47.748623] change   /devices/pci0000:00/0000:00:02.0/drm/card0 (drm)
UDEV  [47.754455] change   /devices/pci0000:00/0000:00:02.0/drm/card0 (drm)
```

2) unplugging DP device from 1
```
KERNEL[62.566260] remove   /devices/pci0000:00/0000:00:02.0/drm/card0/card0-DP-3 (drm)
KERNEL[62.566579] remove   /devices/pci0000:00/0000:00:02.0/drm/card0/card0-DP-4 (drm)
KERNEL[62.566643] remove   /devices/pci0000:00/0000:00:02.0/i2c-10 (i2c)
KERNEL[62.566904] remove   /devices/pci0000:00/0000:00:02.0/drm/card0/card0-DP-5 (drm)
KERNEL[62.567011] change   /devices/pci0000:00/0000:00:02.0/drm/card0 (drm)
UDEV  [62.571670] remove   /devices/pci0000:00/0000:00:02.0/drm/card0/card0-DP-3 (drm)
UDEV  [62.572112] remove   /devices/pci0000:00/0000:00:02.0/drm/card0/card0-DP-4 (drm)
UDEV  [62.572327] remove   /devices/pci0000:00/0000:00:02.0/i2c-10 (i2c)
UDEV  [62.573119] remove   /devices/pci0000:00/0000:00:02.0/drm/card0/card0-DP-5 (drm)
UDEV  [62.574517] change   /devices/pci0000:00/0000:00:02.0/drm/card0 (drm)
KERNEL[62.576553] change   /devices/pci0000:00/0000:00:02.0/drm/card0 (drm)
UDEV  [62.577952] change   /devices/pci0000:00/0000:00:02.0/drm/card0 (drm)
```

3) plugging DP device into different type-C port (host TR, dock SC)
```
KERNEL[166.768286] add      /devices/pci0000:00/0000:00:02.0/drm/card0/card0-DP-3 (drm)
KERNEL[166.768369] change   /devices/pci0000:00/0000:00:02.0/drm/card0 (drm)
KERNEL[166.775536] add      /devices/pci0000:00/0000:00:02.0/i2c-10 (i2c)
KERNEL[166.776410] add      /devices/pci0000:00/0000:00:02.0/drm/card0/card0-DP-4 (drm)
KERNEL[166.776493] change   /devices/pci0000:00/0000:00:02.0/drm/card0 (drm)
UDEV  [166.776571] add      /devices/pci0000:00/0000:00:02.0/drm/card0/card0-DP-3 (drm)
KERNEL[166.776623] add      /devices/pci0000:00/0000:00:02.0/drm/card0/card0-DP-5 (drm)
KERNEL[166.776687] change   /devices/pci0000:00/0000:00:02.0/drm/card0 (drm)
KERNEL[166.776738] change   /devices/pci0000:00/0000:00:02.0/drm/card0 (drm)
UDEV  [166.779462] add      /devices/pci0000:00/0000:00:02.0/i2c-10 (i2c)
UDEV  [166.780330] change   /devices/pci0000:00/0000:00:02.0/drm/card0 (drm)
UDEV  [166.785346] add      /devices/pci0000:00/0000:00:02.0/drm/card0/card0-DP-4 (drm)
UDEV  [166.787076] change   /devices/pci0000:00/0000:00:02.0/drm/card0 (drm)
UDEV  [166.790748] add      /devices/pci0000:00/0000:00:02.0/drm/card0/card0-DP-5 (drm)
UDEV  [166.792866] change   /devices/pci0000:00/0000:00:02.0/drm/card0 (drm)
UDEV  [166.794487] change   /devices/pci0000:00/0000:00:02.0/drm/card0 (drm)
KERNEL[166.796468] change   /devices/pci0000:00/0000:00:02.0/drm/card0 (drm)
KERNEL[166.796566] change   /devices/pci0000:00/0000:00:02.0/drm/card0 (drm)
UDEV  [166.798215] change   /devices/pci0000:00/0000:00:02.0/drm/card0 (drm)
UDEV  [166.799940] change   /devices/pci0000:00/0000:00:02.0/drm/card0 (drm)
KERNEL[168.485259] add      /devices/pci0000:00/0000:00:1d.0/0000:03:00.0/0000:04:02.0/0000:39:00.0/usb4/4-1 (usb)
KERNEL[168.487326] add      /devices/pci0000:00/0000:00:1d.0/0000:03:00.0/0000:04:02.0/0000:39:00.0/usb4/4-1/4-1:1.0 (usb)
KERNEL[168.488741] bind     /devices/pci0000:00/0000:00:1d.0/0000:03:00.0/0000:04:02.0/0000:39:00.0/usb4/4-1/4-1:1.0 (usb)
KERNEL[168.488877] bind     /devices/pci0000:00/0000:00:1d.0/0000:03:00.0/0000:04:02.0/0000:39:00.0/usb4/4-1 (usb)
UDEV  [168.493727] add      /devices/pci0000:00/0000:00:1d.0/0000:03:00.0/0000:04:02.0/0000:39:00.0/usb4/4-1 (usb)
UDEV  [168.501220] add      /devices/pci0000:00/0000:00:1d.0/0000:03:00.0/0000:04:02.0/0000:39:00.0/usb4/4-1/4-1:1.0 (usb)
UDEV  [168.510120] bind     /devices/pci0000:00/0000:00:1d.0/0000:03:00.0/0000:04:02.0/0000:39:00.0/usb4/4-1/4-1:1.0 (usb)
UDEV  [168.517465] bind     /devices/pci0000:00/0000:00:1d.0/0000:03:00.0/0000:04:02.0/0000:39:00.0/usb4/4-1 (usb)
KERNEL[168.792648] add      /devices/pci0000:00/0000:00:1d.0/0000:03:00.0/0000:04:02.0/0000:39:00.0/usb4/4-1/4-1.3 (usb)
KERNEL[168.804325] add      /devices/pci0000:00/0000:00:1d.0/0000:03:00.0/0000:04:02.0/0000:39:00.0/usb4/4-1/4-1.3/4-1.3:1.0 (usb)
KERNEL[168.811157] bind     /devices/pci0000:00/0000:00:1d.0/0000:03:00.0/0000:04:02.0/0000:39:00.0/usb4/4-1/4-1.3/4-1.3:1.0 (usb)
KERNEL[168.811336] bind     /devices/pci0000:00/0000:00:1d.0/0000:03:00.0/0000:04:02.0/0000:39:00.0/usb4/4-1/4-1.3 (usb)
KERNEL[168.888839] add      /devices/pci0000:00/0000:00:1d.0/0000:03:00.0/0000:04:02.0/0000:39:00.0/usb4/4-1/4-1.4 (usb)
KERNEL[168.890332] add      /devices/pci0000:00/0000:00:1d.0/0000:03:00.0/0000:04:02.0/0000:39:00.0/usb4/4-1/4-1.4/4-1.4:2.0 (usb)
KERNEL[168.891821] add      /devices/pci0000:00/0000:00:1d.0/0000:03:00.0/0000:04:02.0/0000:39:00.0/usb4/4-1/4-1.4/4-1.4:2.1 (usb)
KERNEL[168.891900] bind     /devices/pci0000:00/0000:00:1d.0/0000:03:00.0/0000:04:02.0/0000:39:00.0/usb4/4-1/4-1.4 (usb)
KERNEL[168.892201] remove   /devices/pci0000:00/0000:00:1d.0/0000:03:00.0/0000:04:02.0/0000:39:00.0/usb4/4-1/4-1.4/4-1.4:2.0 (usb)
KERNEL[168.892301] remove   /devices/pci0000:00/0000:00:1d.0/0000:03:00.0/0000:04:02.0/0000:39:00.0/usb4/4-1/4-1.4/4-1.4:2.1 (usb)
UDEV  [168.893651] add      /devices/pci0000:00/0000:00:1d.0/0000:03:00.0/0000:04:02.0/0000:39:00.0/usb4/4-1/4-1.3 (usb)
KERNEL[168.897418] add      /devices/pci0000:00/0000:00:1d.0/0000:03:00.0/0000:04:02.0/0000:39:00.0/usb4/4-1/4-1.4/4-1.4:1.0 (usb)
UDEV  [168.901830] add      /devices/pci0000:00/0000:00:1d.0/0000:03:00.0/0000:04:02.0/0000:39:00.0/usb4/4-1/4-1.3/4-1.3:1.0 (usb)
UDEV  [168.907846] bind     /devices/pci0000:00/0000:00:1d.0/0000:03:00.0/0000:04:02.0/0000:39:00.0/usb4/4-1/4-1.3/4-1.3:1.0 (usb)
UDEV  [168.914770] bind     /devices/pci0000:00/0000:00:1d.0/0000:03:00.0/0000:04:02.0/0000:39:00.0/usb4/4-1/4-1.3 (usb)
KERNEL[169.006751] add      /devices/pci0000:00/0000:00:1d.0/0000:03:00.0/0000:04:02.0/0000:39:00.0/usb4/4-1/4-1.4/4-1.4:1.0/net/eth0 (net)
KERNEL[169.006789] add      /devices/pci0000:00/0000:00:1d.0/0000:03:00.0/0000:04:02.0/0000:39:00.0/usb4/4-1/4-1.4/4-1.4:1.0/net/eth0/queues/rx-0 (queues)
KERNEL[169.006810] add      /devices/pci0000:00/0000:00:1d.0/0000:03:00.0/0000:04:02.0/0000:39:00.0/usb4/4-1/4-1.4/4-1.4:1.0/net/eth0/queues/tx-0 (queues)
KERNEL[169.007020] bind     /devices/pci0000:00/0000:00:1d.0/0000:03:00.0/0000:04:02.0/0000:39:00.0/usb4/4-1/4-1.4/4-1.4:1.0 (usb)
UDEV  [169.036548] add      /devices/pci0000:00/0000:00:1d.0/0000:03:00.0/0000:04:02.0/0000:39:00.0/usb4/4-1/4-1.4 (usb)
UDEV  [169.392046] add      /devices/pci0000:00/0000:00:1d.0/0000:03:00.0/0000:04:02.0/0000:39:00.0/usb4/4-1/4-1.4/4-1.4:2.0 (usb)
UDEV  [169.393204] add      /devices/pci0000:00/0000:00:1d.0/0000:03:00.0/0000:04:02.0/0000:39:00.0/usb4/4-1/4-1.4/4-1.4:2.1 (usb)
UDEV  [169.900632] bind     /devices/pci0000:00/0000:00:1d.0/0000:03:00.0/0000:04:02.0/0000:39:00.0/usb4/4-1/4-1.4 (usb)
UDEV  [169.904930] remove   /devices/pci0000:00/0000:00:1d.0/0000:03:00.0/0000:04:02.0/0000:39:00.0/usb4/4-1/4-1.4/4-1.4:2.1 (usb)
UDEV  [169.904996] remove   /devices/pci0000:00/0000:00:1d.0/0000:03:00.0/0000:04:02.0/0000:39:00.0/usb4/4-1/4-1.4/4-1.4:2.0 (usb)
UDEV  [169.907203] add      /devices/pci0000:00/0000:00:1d.0/0000:03:00.0/0000:04:02.0/0000:39:00.0/usb4/4-1/4-1.4/4-1.4:1.0 (usb)
KERNEL[169.912430] move     /devices/pci0000:00/0000:00:1d.0/0000:03:00.0/0000:04:02.0/0000:39:00.0/usb4/4-1/4-1.4/4-1.4:1.0/net/enxd8d0900353ea (net)
UDEV  [169.930058] add      /devices/pci0000:00/0000:00:1d.0/0000:03:00.0/0000:04:02.0/0000:39:00.0/usb4/4-1/4-1.4/4-1.4:1.0/net/enxd8d0900353ea (net)
UDEV  [169.936410] add      /devices/pci0000:00/0000:00:1d.0/0000:03:00.0/0000:04:02.0/0000:39:00.0/usb4/4-1/4-1.4/4-1.4:1.0/net/eth0/queues/tx-0 (queues)
UDEV  [169.936552] add      /devices/pci0000:00/0000:00:1d.0/0000:03:00.0/0000:04:02.0/0000:39:00.0/usb4/4-1/4-1.4/4-1.4:1.0/net/eth0/queues/rx-0 (queues)
UDEV  [169.940496] bind     /devices/pci0000:00/0000:00:1d.0/0000:03:00.0/0000:04:02.0/0000:39:00.0/usb4/4-1/4-1.4/4-1.4:1.0 (usb)
UDEV  [169.946741] move     /devices/pci0000:00/0000:00:1d.0/0000:03:00.0/0000:04:02.0/0000:39:00.0/usb4/4-1/4-1.4/4-1.4:1.0/net/enxd8d0900353ea (net)
KERNEL[177.284359] change   /devices/pci0000:00/0000:00:02.0/drm/card0 (drm)
UDEV  [177.286610] change   /devices/pci0000:00/0000:00:02.0/drm/card0 (drm)
```

4) unplug from that type C in 3
``
KERNEL[198.525942] remove   /devices/pci0000:00/0000:00:02.0/drm/card0/card0-DP-3 (drm)
KERNEL[198.526243] remove   /devices/pci0000:00/0000:00:02.0/drm/card0/card0-DP-4 (drm)
KERNEL[198.526623] remove   /devices/pci0000:00/0000:00:02.0/i2c-10 (i2c)
KERNEL[198.526725] remove   /devices/pci0000:00/0000:00:02.0/drm/card0/card0-DP-5 (drm)
KERNEL[198.526829] change   /devices/pci0000:00/0000:00:02.0/drm/card0 (drm)
UDEV  [198.532056] remove   /devices/pci0000:00/0000:00:02.0/drm/card0/card0-DP-3 (drm)
UDEV  [198.532616] remove   /devices/pci0000:00/0000:00:02.0/drm/card0/card0-DP-4 (drm)
UDEV  [198.532874] remove   /devices/pci0000:00/0000:00:02.0/i2c-10 (i2c)
UDEV  [198.533501] remove   /devices/pci0000:00/0000:00:02.0/drm/card0/card0-DP-5 (drm)
UDEV  [198.534914] change   /devices/pci0000:00/0000:00:02.0/drm/card0 (drm)
KERNEL[198.536902] change   /devices/pci0000:00/0000:00:02.0/drm/card0 (drm)
UDEV  [198.538408] change   /devices/pci0000:00/0000:00:02.0/drm/card0 (drm)
```

5) DP device plugged into TBT dock, using TBT port host
```
KERNEL[308.282372] add      /devices/pci0000:00/0000:00:02.0/i2c-10 (i2c)
KERNEL[308.282920] add      /devices/pci0000:00/0000:00:02.0/drm/card0/card0-DP-3 (drm)
KERNEL[308.282985] change   /devices/pci0000:00/0000:00:02.0/drm/card0 (drm)
KERNEL[308.283064] add      /devices/pci0000:00/0000:00:02.0/drm/card0/card0-DP-4 (drm)
KERNEL[308.283113] change   /devices/pci0000:00/0000:00:02.0/drm/card0 (drm)
KERNEL[308.283192] add      /devices/pci0000:00/0000:00:02.0/drm/card0/card0-DP-5 (drm)
KERNEL[308.283229] change   /devices/pci0000:00/0000:00:02.0/drm/card0 (drm)
KERNEL[308.283263] change   /devices/pci0000:00/0000:00:02.0/drm/card0 (drm)
UDEV  [308.285915] add      /devices/pci0000:00/0000:00:02.0/i2c-10 (i2c)
UDEV  [308.291127] add      /devices/pci0000:00/0000:00:02.0/drm/card0/card0-DP-3 (drm)
UDEV  [308.294103] change   /devices/pci0000:00/0000:00:02.0/drm/card0 (drm)
UDEV  [308.298782] add      /devices/pci0000:00/0000:00:02.0/drm/card0/card0-DP-4 (drm)
UDEV  [308.300851] change   /devices/pci0000:00/0000:00:02.0/drm/card0 (drm)
KERNEL[308.304479] change   /devices/pci0000:00/0000:00:02.0/drm/card0 (drm)
UDEV  [308.304999] add      /devices/pci0000:00/0000:00:02.0/drm/card0/card0-DP-5 (drm)
UDEV  [308.306816] change   /devices/pci0000:00/0000:00:02.0/drm/card0 (drm)
UDEV  [308.308592] change   /devices/pci0000:00/0000:00:02.0/drm/card0 (drm)
UDEV  [308.310175] change   /devices/pci0000:00/0000:00:02.0/drm/card0 (drm)
KERNEL[309.195561] change   /devices/virtual/block/loop0 (block)
UDEV  [309.208630] change   /devices/virtual/block/loop0 (block)
KERNEL[309.225321] change   /devices/virtual/block/loop0 (block)
KERNEL[309.228067] change   /devices/virtual/block/loop0 (block)
UDEV  [309.229863] change   /devices/virtual/block/loop0 (block)
UDEV  [309.232209] change   /devices/virtual/block/loop0 (block)
KERNEL[318.596504] change   /devices/pci0000:00/0000:00:02.0/drm/card0 (drm)
UDEV  [318.603997] change   /devices/pci0000:00/0000:00:02.0/drm/card0 (drm)
```

6) unplug from 5
```
KERNEL[345.688515] remove   /devices/pci0000:00/0000:00:02.0/drm/card0/card0-DP-3 (drm)
KERNEL[345.688637] remove   /devices/pci0000:00/0000:00:02.0/i2c-10 (i2c)
KERNEL[345.688789] remove   /devices/pci0000:00/0000:00:02.0/drm/card0/card0-DP-4 (drm)
KERNEL[345.688920] remove   /devices/pci0000:00/0000:00:02.0/drm/card0/card0-DP-5 (drm)
KERNEL[345.689050] change   /devices/pci0000:00/0000:00:02.0/drm/card0 (drm)
UDEV  [345.694048] remove   /devices/pci0000:00/0000:00:02.0/drm/card0/card0-DP-3 (drm)
UDEV  [345.694321] remove   /devices/pci0000:00/0000:00:02.0/i2c-10 (i2c)
UDEV  [345.695370] remove   /devices/pci0000:00/0000:00:02.0/drm/card0/card0-DP-4 (drm)
UDEV  [345.695451] remove   /devices/pci0000:00/0000:00:02.0/drm/card0/card0-DP-5 (drm)
UDEV  [345.696876] change   /devices/pci0000:00/0000:00:02.0/drm/card0 (drm)
KERNEL[345.698457] change   /devices/pci0000:00/0000:00:02.0/drm/card0 (drm)
UDEV  [345.699955] change   /devices/pci0000:00/0000:00:02.0/drm/card0 (drm)
```

I was hoping that we might see new nodes for the 3, 4, or 5 DP devices but alas no.
```
sys/devices # find -name "drm*"
./pci0000:00/0000:00:02.0/drm
./pci0000:00/0000:00:02.0/drm/card0/card0-eDP-1/drm_dp_aux0
./pci0000:00/0000:00:02.0/drm/card0/card0-DP-2/drm_dp_aux2
./pci0000:00/0000:00:02.0/drm/card0/card0-DP-1/drm_dp_aux1
```

In my example above the monitor was connected to "DP3"
```
/sys/devices/pci0000:00/0000:00:02.0/drm/card0 #  cat card0-DP-3/edid  
��������@LK7D�,x:�բWR�'PT�KqO����/&@�`��00 5����J6HFT44BD7KL
�DELL P2014H
 �8LS
      �%                                                                                                                                                                                                    
```

However the data for the DPCD command came over aux node 2.
```
VMM5331 in Dell dock
  DeviceId:             d83854e8c81f5d47b0b99c8a3b42ae5127c75eb8
  Guid:                 89fec0b6-6b76-5008-b82c-5e5c6c164007 <- MST-panamera-vmm5331-259
  Summary:              Multi Stream Transport controller
  Flags:                updatable|require-ac|registered
  Vendor:               Dell Inc
  Version:              5.04.000
  VersionFormat:        triplet
  InstallDuration:      95
  Created:              2019-08-19
  PhysicalId:           drm_dp_aux2
  FirmwareSizeMin:      524288
  FirmwareSizeMax:      524288
  CustomFlags:          skip-restart,require-ac
  SynapticsMSTRad:      0
  SynapticsMSTAuxNode:  drm_dp_aux2
  SynapticsMSTKind:     DIRECT
  SynapticsMSTLayer:    0
```

I can't find anyway to tell that the topology is currently such that the monitor comes through DPDDC-C (DP-2)

hughsie commented:
I think that's fine; any drm events we can rate limit and rescan **all** the rm_dp_aux_dev the plugin owns. I'll try that next week.

hughsie commented:
> I'll try that next week.

It actually works wonderfully well using the drm events for the `->rescan()` -- I had to stick a pretty chunky timeout to wait for the link training, but it seems to work with the Lenovo docks (TBT and non-TBT) I've got here. Obviously when/if the kernel sends a uevent for `drm_dp_aux_dev` we can use that instead of the timeout, but for now I need this to work on old versions of the kernel for obvious reasons.

I've had a shot at making cascade actually work too, although I don't actually think the old plugin would work with a cascade device. The new one might :)

superm1 commented:
>It actually works wonderfully well using the drm events for the ->rescan() -- I had to stick a pretty chunky timeout to wait for the link training, but it seems to work with the Lenovo docks (TBT and non-TBT) I've got here. Obviously when/if the kernel sends a uevent for drm_dp_aux_dev we can use that instead of the timeout, but for now I need this to work on old versions of the kernel for obvious reasons.

Well that's good to hear that it works on those.  Have you checked on the Dell ones too?  I don't see a reason it shouldn't work there.

> I've had a shot at making cascade actually work too, although I don't actually think the old plugin would work with a cascade device. The new one might :)

It definitely did work back when Synaptics first did the plugin, and i've checked it from time to time.  It's still possible it did break of course.

hughsie commented:
>  Have you checked on the Dell ones too

No hardware here yet :/

superm1 commented:
> No hardware here yet :/

At least with the "old" pre-production stuff you have you should be able to blacklist loading `dell_dock` and just use `synapticsmst` directly when monitors are plugged in IIRC.  The MST chip doesn't change between hardware generations so it will be able to take a current MST firmware.

hughsie commented:
> but when the dell plugin turns off the GPIO this of course causes a rescan of the bus#

Right, so this is behaving exactly as designed. Perhaps we need another quirk for the FuDevice that is something like `persistent` that doesn't get autoremoved when the DRM device goes away?

hughsie commented:
Can you try ^^ and see if that kind of thing could work?

hughsie commented:
>  the force power and plugins working together is even worth it or we should just axe it all together

I was thinking the same thing the other day, but I do think the force power stuff is useful for the "evil maid" attack, or more likely when you plug your laptop in at a conference into some random projector that lights up the MST hub for the first time. The unlikely scenario is a malicious projector, the much more likely scenario is the MST hub just not working when you need it the most.

As a general note, I think that devices that fall off the bus when there's nothing connected (also looking at you, Thunderbolt) are a huge pain in the ass, and an anti-pattern we want to dissuade people from using. It's really poor-mans sleep state management.

> If you unplug the dock and move to a different port I think you might end up with 2x the devices

I don't think many machines have more than one thunderbolt port, and I don't think it's _likely_ that users would connect to a different port; at least in my case I always connect the tbt adaptor on one side of my desk rather than other.

hughsie commented:
> Machines with updatable MST hubs in them are rarer

Valid point. I think TBT controllers are still valid to force power.

> Fortunately this behavior goes away when machines adopt "Native Thunderbolt"

That is a relief, thanks for sharing that.

> The one I'm using right now has 3 type-C ports, 2 of them Thunderbolt

Too much Lenovo for me then :) I agree for synaptics mst we can probably dump the force power stuff and revert the last patch on this PR. As you're more familiar with the lifecycle of the mst force power stuff, can you add a commit to this branch removing that beahviour please? That also lets us nuke the long g_usleep in fu_plugin_update() too I think.

hughsie commented:
> board ID mismatch, got 0x0301, expected 0x0103

Which is correct?

hughsie commented:
Can you try ^^ please.

hughsie commented:
Okay, fixed up (doh), removed the last FIXME and rebased. Fingers crossed.

hughsie commented:
Ahh, good catch. ^^

hughsie commented:
> come back much later

as in several seconds? I think a good idea about not rescanning on update. ^^

hughsie commented:
I think I'm a lot more comfortable with the new code to the old code. What do you think about merging?

hughsie commented:
Well done on finding that `FWUPD_DEVICE_FLAG_WAIT_FOR_REPLUG` bug; I merged that right to master after adding some tests. I've rebased and squashed down all the patches, I hope that's okay. I just tested on a ThinkPad Thunderbolt Dock I have here and it worked perfectly, and correctly reported the successful update.
